SharedArray documentation (#196)
loadimpact/k6-docs: src/data/markdown/docs/05 Examples/01 Examples/05 data-parameterization.md

<Blockquote mod="warning">

Each VU in k6 is a separate JS VM so in order to not have multiple copies of the whole data file

> [SharedArray](/javascript-api/k6-data/sharedarray) was added. It does have overhead in accessing elements compared to a normal non shared
> array, but the difference is negligible compared to the time it takes to make requests. And is
> especially not comparable as not using it with big files basically means k6 will be using too much
> memory to run. Which might lead to your script not being able to run at all or aborting in the
> middle.

Reviewer: ```suggestion
[SharedArray](/javascript-api/k6-data/sharedarray) was added. It does have some CPU overhead in accessing elements compared to a normal non shared
array, but the difference is negligible compared to the time it takes to make requests. This becomes
even less of an issue compared to not using it with large files, as k6 would otherwise use too much memory to run, which might lead to your script not being able to run at all or aborting in the middle if the system resources are exhausted.
```

> For example, the Cloud service allots 8GB of memory for every 300VUs. So if you are files are big
> enough and you are not using [SharedArray](/javascript-api/k6-data/sharedarray), that might mean that your script will run out of memory at
> some point. Additionally even if there is enough memory, k6 has a garbage collector(as it's written
> in golang) and it will walk through all accessible objects (including JS ones) and figure out which
> need to be garbage collected. For big JS arrays copied hundreds of times this adds quite a lot of
> additional work.

Reviewer: The second sentence just repeats what the last sentence of the previous paragraph said, and mentioning the garbage collector seems like too much detail for this documentation.

Mentioning the Cloud limits might be relevant, but we'll have to keep it updated if that changes, and I'm not sure if it's worth bringing it up at all or if this is the right place for it. So I would scrap this paragraph entirely, but if not some minor suggestions:

```suggestion
For example, the Cloud service allocates 8GB of memory for every 300 VUs. So if your files are large
enough and you are not using [SharedArray](/javascript-api/k6-data/sharedarray), that might mean that your script will run out of memory at
some point. Additionally even if there is enough memory, k6 has a garbage collector (as it's written
in golang) and it will walk through all accessible objects (including JS ones) and figure out which
need to be garbage collected. For big JS arrays copied hundreds of times this adds quite a lot of
additional work.
```

Author: The cloud limits were here from before. I am not certain this is the correct place for them as well, but it is definetely good to have them somewhere. Given that if a cloud user runs out of memory they might be directed to this page, I am for this actually staying. It also shows how much memory we have decided to allocate to a given number of VUs.

As to change ... things changes, so we will need to update it if we ever  change it in the cloud :)

> A note on performance characteristics of `SharedArray` can be found within it's [API documentation](/javascript-api/k6/data/sharedarray#performance-characteristics).

Reviewer: ```suggestion
A note on performance characteristics of `SharedArray` can be found within its [API documentation](/javascript-api/k6/data/sharedarray#performance-characteristics).
```

<CodeGroup labels={["parse-json.js"]} lineNumbers={[true]}>

```javascript
import { SharedArray } from "k6/data";

> // not using ShareArray here will mean that the code in the function call (that is what loads and

Reviewer: ```suggestion
// not using SharedArray here will mean that the code in the function call (that is what loads and
```

```javascript
import papaparse from './papaparse.js';
import { SharedArray } from "k6/data";

> // not using ShareArray here will mean that the code in the function call (that is what loads and

Reviewer: ```suggestion
// not using SharedArray here will mean that the code in the function call (that is what loads and
```

```javascript
import papaparse from 'https://jslib.k6.io/papaparse/5.1.1/index.js';
import { SharedArray } from "k6/data";

> // not using ShareArray here will mean that the code in the function call (that is what loads and

Reviewer: ```suggestion
// not using SharedArray here will mean that the code in the function call (that is what loads and
```

```javascript
/*  Where contents of data.csv is:
username,password
admin,123
test_user,1234
*/
import http from 'k6/http';
import { check, sleep } from 'k6';
import { SharedArray } from 'k6/data'
import papaparse from 'https://jslib.k6.io/papaparse/5.1.1/index.js';

> // not using ShareArray here will mean that the code in the function call (that is what loads and

Reviewer: ```suggestion
// not using SharedArray here will mean that the code in the function call (that is what loads and
```

<br/>

## Generating data

See [this example project on GitHub](https://github.com/k6io/example-data-generation) showing how to use faker.js to generate realistic data at runtime.

> ## Old hacks

Reviewer: :)

```suggestion
## Old workarounds
```
